File: 1-moderate/double-squares/main.cpp
#include <iostream>
#include <sstream>
#include <fstream>
#include <math.h>

unsigned int countDoubleSquares(unsigned int x)
{
    if (x == 0) {
        return 1;
    }

    unsigned int result = 0;

    for (unsigned int b = 0; b <= pow(x, 0.5); b++) {
        unsigned int a = pow(x - pow(b, 2), 0.5);

        if (a < b) {
            break;
        }

        if (x == (pow(a ,2) + pow(b, 2))) {
            result++;
        }
    }

    return result;
}

int main(int argc, char** argv)
{
    std::ifstream file(argv[1]);

    std::string line;
    bool first = true;
    while (std::getline(file, line))
    {
        unsigned int x;   
        std::stringstream ss(line);
        ss >> x;

        if (first) {
            // Skip the number of lines count.
            first = false;
            continue;
        }

        std::cout << countDoubleSquares(x) << std::endl;
    }
}

File: 1-moderate/double-squares/main.py
import sys

def count_double_squares(x):
    if x == 0: return 1
    result = 0
    for b in range(0, int(x ** 0.5)+1):
        a = int((x - (b ** 2)) ** 0.5)
        if a < b: break
        if x == a ** 2 + b ** 2: result += 1
    return result

test_cases = open(sys.argv[1], 'r')
first = True

for test in test_cases:
    test = test.strip()
    if len(test) == 0: continue
    if first:
        first = False
        continue
    x = int(test)
    print(count_double_squares(x))

test_cases.close()
